wgsl/index.bs

```diff
@@ -9133,6 +9133,170 @@ Cube faces are square, so the x and y components of the result are equal.
+* Find the four texels that would be used in a sampling operation with linear filtering,
+    from [=mip level=] 0:
+    * Use the specified coordinate, array index (when present), and offset (when present).
+    * The texels are adjacent, forming a square, when considering their texture space coordinates (*u*,*v*).
+    * Selected texels at the texture edge, cube face edge, or cube corners are handled
+        as in ordinary texture sampling.
+* For each texel, read one channel and convert it into a scalar value.
+    * For non-depth textures, a zero-based `component` parameter specifies the channel to use.
+        * If the texture format supports the specified channel, i.e. has more than `component` channels:
+             * Yield scalar value `v[component]` when the texel value is `v`.
+        * Otherwise:
+             * Yield 0.0 when `component` is 1 or 2.
+             * Yield 1.0 when `component` is 3 (the alpha channel).
+    * For depth textures, yield the texel value. (Depth textures only have one channel.)
+* Yield the four-element vector, arranging scalars produced by the previous step into components
+    according to the relative coordinates of the texels, as follows:
+    * <table>
+        <thead class='data'><td>Result component<td>Relative texel coordinate</thead>
+        <tr><td>x<td>(*u*<sub>min</sub>,*v*<sub>max</sub>)
+        <tr><td>y<td>(*u*<sub>max</sub>,*v*<sub>max</sub>)
+        <tr><td>z<td>(*u*<sub>max</sub>,*v*<sub>min</sub>)
+        <tr><td>w<td>(*u*<sub>min</sub>,*v*<sub>min</sub>)
+        </table>
+
+TODO: The four texels are the "sample footprint" that should be described by the WebGPU spec.
+https://github.com/gpuweb/gpuweb/issues/2343
+
+```rust
+textureGather(t: texture_2d<T>, s: sampler, coords: vec2<f32>) -> vec4<T>
+textureGather(t: texture_2d<T>, s: sampler, coords: vec2<f32>, offset: vec2<i32>) -> vec4<T>
+textureGather(t: texture_2d_array<T>, s: sampler, coords: vec2<f32>, array_index: i32) -> vec4<T>
+textureGather(t: texture_2d_array<T>, s: sampler, coords: vec2<f32>, array_index: i32, offset: vec2<i32>) -> vec4<T>
+textureGather(t: texture_cube<T>, s: sampler, coords: vec3<f32>) -> vec4<T>
+textureGather(t: texture_cube_array<T>, s: sampler, coords: vec3<f32>, array_index: i32) -> vec4<T>
+textureGather(t: texture_depth_2d, s: sampler, coords: vec2<f32>) -> vec4<f32>
+textureGather(t: texture_depth_2d, s: sampler, coords: vec2<f32>, offset: vec2<i32>) -> vec4<f32>
+textureGather(t: texture_depth_2d_array, s: sampler, coords: vec2<f32>, array_index: i32) -> vec4<f32>
+textureGather(t: texture_depth_2d_array, s: sampler, coords: vec2<f32>, array_index: i32, offset: vec2<i32>) -> vec4<f32>
+textureGather(t: texture_depth_cube, s: sampler, coords: vec3<f32>) -> vec4<f32>
+textureGather(t: texture_depth_cube_array, s: sampler, coords: vec3<f32>, array_index: i32) -> vec4<f32>
+textureGatherComponent(t: texture_2d<T>, s: sampler, coords: vec2<f32>, component: i32) -> vec4<T>
```

Comment:
Did you discuss the other choices here with the group? I think, `component` is slightly worse than what other languages have.
Could we go crazy and do something like `textureGather<C=0>(...)`, i.e. a generic parameter defaulting to 0 (Red).
Even though we don't have functions called generically, we have type constructors that very much act like functions, e.g. `vec4<f32>(1.0)`.

Just throwing it out there in case this hasn't been mentioned before.
I think proceeding with PR is ok (minus the other review notes) :)

Reply:
> Did you discuss the other choices here with the group? I think, component is slightly worse than what other languages have.

No.  I'm looking for feedback.  :-)
But I wanted something concrete to discuss.

> Could we go crazy and do something like textureGather<C=0>(...), i.e. a generic parameter defaulting to 0 (Red).

Named parameters would be nicer. #1378



Looking at the combinations, we could add component as an optional last parameter to the existing 'textureGather' functions. The only case where you end up with two adjacent 'i32' parameters is with the forms that use an arrayed texture, and *without* the offset.    (The offsets occur only for 2d and 2d_array, and are 2-dimensional)

          textureGather(t:texture_2d_array, s:sampler, coords:vec2<f32>, array_index:i32, component: i32)

Reply:
Discussed in the [2021-11-30 meeting.](https://docs.google.com/document/d/1dVE2hu9oYJvy34ASYnBf5B8fRHtq1Njy52k6Oi9N1uQ/edit#)